ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1658,6 +1751,133 @@ private void onInvalidRange(Exception exception) {
+      BlobProperties receivedBlobProperties = MessageFormatRecord.deserializeBlobProperties(payload);
+      /*
+       * Some cached variables are mutable.
+       * Update them and then compare if other immutable variables are same or not.
+       */
+      updateTtlIfRequired(serverBlobProperties, messageInfo);
+      lifeVersion = messageInfo.getLifeVersion();
+      blobInfo.setLifeVersion(lifeVersion);
+      if (!serverBlobProperties.equals(receivedBlobProperties)) {
+        /*
+         * This is a rare error case.
+         * If we find blob properties is in our frontend cache, we request just blobInfo from backend to validate
+         * the cache entry. If cached blob properties does not match received blob properties, due to a random bit flip for eg.,
+         * we must delete the cache entry and throw a retriable exception to the user.
+         * The next request will re-fill the cache with a valid entry if the blob stills exists.
+         * We do not need to decrypt anything at this point.
+         */
+        String reason = "Cached blob property does not match received blob property";
+        logger.error("{} for blobId = {}, cached blob property = {}, received blob property = {}", reason, blobId,
+            serverBlobProperties, receivedBlobProperties);
+        deleteMetadata(reason);
+        onInvalidCacheEntry(new RouterException(reason, RouterErrorCode.UnexpectedInternalError));
+        return;
+      }
+      getOptions().ageAtAccessTracker.trackAgeAtAccess(serverBlobProperties.getCreationTimeInMs());
+
+      /*
+       * If we reach here, then the cached metadata is valid and already decrypted from a previous GET/RANGE request.
+       * Just use it directly and initiate fetching data chunks.
+       */
+      prepareDataChunks();
+    }
+
+    /**
+     * Resolve range, if needed, and request data chunks
+     * @throws IllegalArgumentException
+     */
+    void prepareDataChunks() throws IllegalArgumentException {
+      long rangeTotalSize = totalSize;
+      if (options.getBlobOptions.hasBlobSegmentIdx()) {
+        int requestedSegment = options.getBlobOptions.getBlobSegmentIdx();
+        if (requestedSegment < 0 || requestedSegment >= chunkMetadataList.size()) {
+          throw new IllegalArgumentException(
+              "Bad segment number: " + requestedSegment + ", num of keys: " + chunkMetadataList.size());
+        }
+        chunkMetadataList = chunkMetadataList.subList(requestedSegment, requestedSegment + 1);
+        rangeTotalSize = chunkMetadataList.get(0).getSize();
+      }
+      if (options.getBlobOptions.getRange() != null && resolveRangeProper(rangeTotalSize)) {
+        if (!options.getBlobOptions.hasBlobSegmentIdx()) {
+          chunkMetadataList = compositeBlobInfo.getStoreKeysInByteRange(resolvedByteRange.getStartOffset(),
+              resolvedByteRange.getEndOffset());
+        }
+        initializeDataChunks();
```

Comment:
Would you need to call `initializeDataChunks` outside the `if (options.getBlobOptions.getRange() != null)` since we also need to initialize it for segmented GETs now. But you might still need to only initialize if range resolution passed before initializing.

Reply:
problem with segmented get
fixBlobSizeIfRequired(blobInfo.getBlobProperties().getBlobSize(),
                  options.getBlobOptions.hasBlobSegmentIdx() ? firstChunk.getChunkMetadataList().get(0).getSize()
                      : totalSize);

This changes the size of the blob in the metadata cache and then the cached entry doesn't match blobinfo coming from backend. Ignoring SEGMENTED GET as discussed offline